### PR TITLE
Fix MayUseToken vulnerability in o1js

### PR DESCRIPTION
**Pull request by MartinMinkov**
o1js: https://github.com/o1-labs/o1js/pull/1741
bindings: https://github.com/o1-labs/o1js-bindings/pull/280

## Overview
This PR adds a needs_custom_js annotation to the MayUseToken type in our OCaml code. This change allows us to define custom constraints in the generated TypeScript/JavaScript code, specifically to ensure that the `parentsOwnToken` and `inheritFromParent` fields cannot both be true simultaneously.

## Changes
- Added needs_custom_js annotation to the MayUseToken type deriver.

## Motivation
The primary goal is to enhance type safety and prevent potential vulnerabilities by allowing more precise control over the TypeScript representation of our OCaml types. This change enables us to implement custom check methods in o1js, ensuring well-formedness when proving circuits.

## Technical Details
The `needs_custom_js` annotation is applied with a custom JS type layout.
The layout specifies two boolean fields: `parentsOwnToken` and `inheritFromParent`.
This change doesn't alter the OCaml type structure but provides a hook for custom JS/TS type definitions.

## Implementation Notes
The actual implementation of the custom `check` method in o1js will be done in a separate PR.
This change only modifies the type derivation, not the actual data structure or its usage.

### CODE REVIEW

File: src/lib/mina_base/account_update.ml [hunk 1]
```diff
@@ -1137,14 +1137,23 @@ module Body = struct
...
       let open Fields_derivers_zkapps in
+      let may_use_token =
+        needs_custom_js
+          ~js_type:
+            (js_record
+               [ ("parentsOwnToken", js_layout bool)
+               ; ("inheritFromParent", js_layout bool)
+               ] )
+          ~name:"MayUseToken" May_use_token.deriver
+      in
```
Comment (mitschabaude) [inline src/lib/mina_base/account_update.ml:1148]: nit: it might be a bit cleaner to add this logic to `May_use_token.deriver`, not here

**Review by mitschabaude — APPROVED** [review]

File: src/lib/mina_base/account_update.ml [hunk 2]
```diff
@@ -477,8 +477,17 @@ module May_use_token = struct
 
   let deriver obj =
     let open Fields_derivers_zkapps in
-    iso_record ~of_record:As_record.to_variant ~to_record:As_record.of_variant
-      As_record.deriver obj
+    let may_use_token =
+      iso_record ~of_record:As_record.to_variant ~to_record:As_record.of_variant
+        As_record.deriver
+    in
+    needs_custom_js
```
Comment (georgeee) [inline src/lib/mina_base/account_update.ml:484]: Could you add some information to PR description:

- What was the format of `may_use_token` field before this PR
- What will be after this PR
- What's motivation of changing the format?

**Review by georgeee — APPROVED** [review]

File: src/lib/mina_base/account_update.ml (hunk 2 above)
Comment (MartinMinkov) [reply src/lib/mina_base/account_update.ml:484]: done!

**Comment by MartinMinkov** [comment]
!ci-build-me

**Pull request merged**